network/dnstruncate/packet_proxy.go

```diff
@@ -0,0 +1,154 @@
+
+// dnsTruncateProxy is a network.PacketProxy that create dnsTruncateRequestHandler to handle DNS requests locally.
+//
+// Multiple goroutines may invoke methods on a dnsTruncateProxy simultaneously.
+type dnsTruncateProxy struct {
+}
+
+// dnsTruncateRequestHandler is a network.PacketRequestSender that handles DNS requests in UDP protocol locally,
+// without sending the requests to the actual DNS resolver. It sets the TC (truncated) bit in the DNS response header
+// to tell the caller to resend the DNS request over TCP.
+//
+// Multiple goroutines may invoke methods on a dnsTruncateProxy simultaneously.
+type dnsTruncateRequestHandler struct {
+	closed     atomic.Bool
+	respWriter network.PacketResponseReceiver
+}
+
+// Compilation guard against interface implementation
+var _ network.PacketProxy = (*dnsTruncateProxy)(nil)
+var _ network.PacketRequestSender = (*dnsTruncateRequestHandler)(nil)
+
+// NewPacketProxy creates a new [network.PacketProxy] that can be used to handle DNS requests if the remote proxy
+// doesn't support UDP traffic. It sets the TC (truncated) bit in the DNS response header to tell the caller to resend
+// the DNS request over TCP.
+//
+// This [network.PacketProxy] should only be used if the remote proxy server doesn't support UDP traffic at all. Note
+// that all other non-DNS UDP packets will be dropped by this [network.PacketProxy].
+func NewPacketProxy() (network.PacketProxy, error) {
+	return &dnsTruncateProxy{}, nil
+}
+
+// NewSession implements [network.PacketProxy].NewSession(). It creates a new [network.PacketRequestSender] that will
+// set the TC (truncated) bit and write the response to `respWriter`.
+func (p *dnsTruncateProxy) NewSession(respWriter network.PacketResponseReceiver) (network.PacketRequestSender, error) {
+	if respWriter == nil {
+		return nil, errors.New("respWriter is required")
+	}
+	return &dnsTruncateRequestHandler{
+		respWriter: respWriter,
+	}, nil
+}
+
+// Close implements [network.PacketRequestSender].Close(), and it closes the corresponding
+// [network.PacketResponseReceiver].
+func (h *dnsTruncateRequestHandler) Close() error {
+	if !h.closed.CompareAndSwap(false, true) {
+		return network.ErrClosed
+	}
+	h.respWriter.Close()
+	return nil
+}
+
+// WriteTo implements [network.PacketRequestSender].WriteTo(). It parses a packet from p, and determines whether it is
+// a valid DNS request. If so, it will write the DNS response with TC (truncated) bit set to the corresponding
+// [network.PacketResponseReceiver] passed to NewSession. If it is not a valid DNS request, the packet will be
+// discarded and returns an error.
+func (h *dnsTruncateRequestHandler) WriteTo(p []byte, destination netip.AddrPort) (int, error) {
+	if h.closed.Load() {
+		return 0, network.ErrClosed
+	}
+	if destination.Port() != standardDNSPort {
+		return 0, fmt.Errorf("non-DNS UDP is not supported: %v %w", destination.Port(), network.ErrPortUnreachable)
```

Comment:
I believe this will read better, and is better aligned with the test (we really just test the port here, not the protocol)
```suggestion
		return 0, fmt.Errorf("UDP traffic to non-DNS port %v is not supported: %w", destination.Port(), network.ErrPortUnreachable)
```